From 7662ce6bcda9f153b164ef57fdadcc5ce6a21088 Mon Sep 17 00:00:00 2001
From: Damiano galassi <damiog@gmail.com>
Date: Tue, 28 Feb 2023 11:23:18 +0100
Subject: [PATCH 3/3] movenc: write 3gpp track names tags for all  available 
 languages

Metadata keys can have a language suffix.  Iterate through all language
variations of the metadata key.

Original patch by John Stebbins <jstebbins@jetheaddev.com>.
---
 libavformat/movenc.c | 25 +++++++++++++++++++++++--
 1 file changed, 23 insertions(+), 2 deletions(-)

diff --git a/libavformat/movenc.c b/libavformat/movenc.c
index cba6ce1..3e680c4 100644
--- a/libavformat/movenc.c
+++ b/libavformat/movenc.c
@@ -28,6 +28,7 @@
 
 #include "movenc.h"
 #include "avformat.h"
+#include "avlanguage.h"
 #include "avio_internal.h"
 #include "dovi_isom.h"
 #include "riff.h"
@@ -3586,7 +3587,14 @@ static int mov_write_3gp_udta_tag(AVIOContext *pb, AVDictionary *metadata,
     if (!strcmp(tag, "yrrc"))
         avio_wb16(pb, atoi(t->value));
     else {
-        avio_wb16(pb, language_code("eng")); /* language */
+        int lang = 0, len;
+        len = strlen(t->key);
+        if (t->key[len - 4] == '-') {
+            lang = ff_mov_iso639_to_lang(&t->key[len - 3], 1);
+        }
+        if (!lang)
+            lang = ff_mov_iso639_to_lang("und", 1);
+        avio_wb16(pb, lang); /* language */
         avio_write(pb, t->value, strlen(t->value) + 1); /* UTF8 string value */
         if (!strcmp(tag, "albm") &&
             (t = av_dict_get(metadata, "track", NULL, 0)))
@@ -3670,8 +3678,21 @@ static int mov_write_track_udta_tag(AVIOContext *pb, MOVMuxContext *mov,
         return ret;
 
     if (mov->mode & (MODE_MP4|MODE_MOV)) {
+        AVDictionaryEntry *t = NULL;
+        int und = 0;
+
         mov_write_track_metadata(pb_buf, st, "name", "title");
-        mov_write_3gp_udta_tag(pb_buf, st->metadata, "titl", "title");
+        while ((t = av_dict_get(st->metadata, "title-", t, AV_DICT_IGNORE_SUFFIX))) {
+            int len = strlen(t->key);
+            if (len == 10 &&
+                ff_convert_lang_to(&t->key[len - 3], AV_LANG_ISO639_2_BIBL)) {
+                mov_write_3gp_udta_tag(pb_buf, st->metadata, "titl", t->key);
+                if (!strcmp("und", &t->key[len - 3]))
+                    und = 1;
+            }
+        }
+        if (!und)
+            mov_write_3gp_udta_tag(pb_buf, st->metadata, "titl", "title");
     }
 
     if (mov->mode & MODE_MP4) {
-- 
2.37.1 (Apple Git-137.1)

